### PR TITLE
Add buildAsync method for async graph compilation

### LINKED ISSUE #263: Support asynchronous graph compilation
Feedback from @wacky6 and @huningxin. Detail is in PR #257.

### PR DESCRIPTION
**Pull request by huningxin**
fix #263

@wchao1115 @anssiko @RafaelCintron @wacky6 PTAL, thanks!

***
<a href="https://pr-preview.s3.amazonaws.com/huningxin/webnn/pull/266.html" title="Last updated on May 7, 2022, 8:53 AM UTC (ac146e9)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/webmachinelearning/webnn/266/4c5b70b...huningxin:ac146e9.html" title="Last updated on May 7, 2022, 8:53 AM UTC (ac146e9)">Diff</a>

### CODE REVIEW

File: index.bs [hunk 1]
```diff
@@ -480,10 +480,11 @@ that shares the same buffer as the input tensor. (In the case of reshape or sque
 the entire data is shared, while in the case of slice, a part of the input data is shared.)
 The implementation may use views, as above, for intermediate values.
 
-The {{MLGraphBuilder}}.{{MLGraphBuilder/build()}} method of the {{MLGraphBuilder}} interface is used to compile and optimize
-the computation graph used to compute one or more specified outputs. The key
-purpose of the compilation step is to enable optimizations that span two or
-more operations, such as operation or loop fusion.
+Before the execution, the computation graph that is used to compute one or more specified outputs needs to be compiled and optimized. The key purpose of the compilation step is to enable optimizations that span two or more operations, such as operation or loop fusion.
+
+There are multiple ways by which the graph may be compiled. The {{MLGraphBuilder}}.{{MLGraphBuilder/build()}} method represents a way the compilation of the graph is carried out immediately on the calling thread, which must also be a worker thread, either on a CPU or GPU device. The compilation produces the {{MLGraph}} that represents a compiled graph for optimal execution.
```
Comment (wacky6) [inline index.bs:485]: WDYT: 

```
The {{MLGraphBuilder}}.{{MLGraphBuilder/build()}} method compiles the graph immediately on the calling thread, which must be a worker thread running on CPU or GPU device.
```

File: index.bs [hunk 2]
```diff
@@ -480,10 +480,11 @@ that shares the same buffer as the input tensor. (In the case of reshape or sque
...
+
+The {{MLGraphBuilder}}.{{MLGraphBuilder/buildAsync()}} method represents a way the compilation of the graph is performed asynchronously either on a parallel timeline in a separate worker thread for the CPU execution or on a GPU timeline in a GPU command queue. This method returns immediately without blocking the calling thread while the actual compilation is offloaded to a different timeline. This type of compilation is appropriate when the responsiveness of the calling thread is critical to good user experience. The {{MLGraph}} will be returned at the time the compilation is successfully completed on the offloaded timeline at which time the calling thread is signaled.
```
Comment (wacky6) [inline index.bs:487]: WDYT: 

buildAsync() compiles the graph in background without blocking the calling thread, and returns a {{Promise}} that resolves to {{MLGraph}}.

File: index.bs [hunk 3]
```diff
@@ -986,14 +987,17 @@ interface MLGraphBuilder {
...
+Both {{MLGraphBuilder}}.{{MLGraphBuilder/build()}} and {{MLGraphBuilder}}.{{MLGraphBuilder/buildAsync()}} methods compile the graph builder state up to the specified output operands into a compiled graph according to the type of {{MLContext}} that creates it. Since this operation can be costly in some machine configurations, the calling thread of the {{MLGraphBuilder}}.{{MLGraphBuilder/build()}} method must only be a worker thread to avoid potential disruption of the user experience. When the {{[[contextType]]}} of the {{MLContext}} is set to [=default-context|default=], the compiled graph is initialized right before the {{MLGraph}} returns. This graph initialization stage is important for optimal performance of the subsequent graph executions. See [[#api-mlcommandencoder-graph-initialization]] for more detail.
```
Comment (wacky6) [inline index.bs:1000]: Perhaps:

"right before the {{MLGraph}} returns"
->
"right before the {{MLGraph}} *is returned*"

**Review by wacky6 — COMMENTED** [review]
lgtm / a few suggestions on phrasing

**Comment by huningxin** [comment]
Thanks @wacky6 for review and suggestion. I've fixed those in the new commit https://github.com/webmachinelearning/webnn/pull/266/commits/ac146e9e085ed848c6987ee39085071cf7a84aae. Please take another look.

**Review by wacky6 — APPROVED** [review]
lgtm

**Review by anssiko — APPROVED** [review]
Thank you @huningxin, for discussion on today's call.

Related, for this method and others, we should add the method steps per https://github.com/webmachinelearning/webnn/issues/210. I won't block this PR on that because we track that separately.

**Review by wchao1115 — APPROVED** [review]

**Comment by huningxin** [comment]
> Related, for this method and others, we should add the method steps per #210. I won't block this PR on that because we track that separately.

Acknowledged that, thanks @anssiko .

And thanks all for review and approvals. With that, I am going to merge it.

**Pull request merged**